### PR TITLE
OpenSSL switch, and finally a robust version.

### PR DESCRIPTION
**Pull request by BenWiederhake**
- avoids OpenSSL stuff (no OpenSSL, not in telegram-purple and not in tgl)
- fixes the "double .insert" issue (see https://github.com/vysheng/tgl/issues/49 )
- use the new `tglpub` format

I'm pretty confident that this version should be robust, as in: No significant new bugs.
(I bet that it breaks as soon as I hit enter ...)

### CODE REVIEW

**Comment by BenWiederhake** [comment]
Closes: #1

**Comment by Croydon** [comment]
Are there more reasons to move away from OpenSSL then the license? The license is going to change to Apache v2 (which I mentioned now multiple times).

**Comment by BenWiederhake** [comment]
And just like you have already been told, APL and GPLv2 are incompatible. Furthermore, the easiest way to sidestep this issue is by simply using GPL'd software.

However, that's just parroting what other people wrote. Let me try again.

https://www.apache.org/licenses/GPL-compatibility.html Here the Apache people tell you that their license is incompatible with the GPLv2:

    Despite our best efforts, the FSF has never considered the Apache License to be
    compatible with GPL version 2, citing the patent termination and indemnification
    provisions as restrictions not present in the older GPL license. The Apache Software
    Foundation believes that you should always try to obey the constraints expressed by
    the copyright holder when redistributing their work.

So yeah, no, still not possible.

In a way, the new license makes it *worse*. With the old license, the issue was just one ambiguous paragraph, which was only interpreted by Debian as a violation. The new license is a definite violation.

So hooray, we got out of that before BadThings:tm: happened.

On an unrelated note, I find both arguments of incompatibility worse than unconvincing. However, "legally binding" and "utterly childish" aren't mutually exclusive, so we have to deal with it.

**Comment by BenWiederhake** [comment]
Oh and btw: The LibreSSL people will probably have lots of arguments about why OpenSSL is bad. But I can't really answer your question, for I have far too little experience working with OpenSSL.

**Comment by majn** [comment]
@Croydon The problem with the license switch is that we do not know when or if its going to happen. As far as I know they merely have declared that they _want_ to switch in that blog post. They need permission from every OpenSSL dev, so it might take years if its even possible at all. I don't think that this is anything we can wait on.

**Comment by majn** [comment]
@BenWiederhake The page you linked says that its possible to include Apache software into GPLv3 projects but just not the other way around. As all parts and libpurple are GPLv2+, using OpenSSL might be legal at some point in the future (if the switch works).

However, I don't think that there is a point in discussing this anymore. The effort of porting to libgcrypt is already done and we can finally get rid of all licensing problems right now not eventually at some undefined point in the future :)

**Comment by EionRobb** [comment]
Nice work @BenWiederhake :)

**Comment by EionRobb** [comment]
In telegram-base.c you should return TRUE or FALSE rather than 1 or 0 in the assert function that now returns a gboolean. You could also potentially use the sha1 cipher already part of libpurple so that you don't need gcrypt included. 

Is it possible to use libpurple ssl funcs to load the pem and convert to the new format? I believe (but am probably mistaken) that the OTR plugin might do this, since they also use gcrypt and pems?

**Comment by BenWiederhake** [comment]
@EionRobb: Your comment about TRUE and FALSE is correct, but I wanted to follow the libtgl convention, as it uses 0 and 1 instead.

Regarding gcrypt: We would still rely on gcrypt, because libtgl (with `./configure --disable-openssl`) now does. So that's not an extra dependency.
